### PR TITLE
gh-122561: Clean up and microoptimize str.translate and charmap codec

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
* Replace PyLong_AS_LONG() with PyLong_AsLong().
* Call PyLong_AsLong() only once per the replacement code.
* Use PyMapping_GetOptionalItem() instead of PyObject_GetItem().

* Issue: gh-122561

### CODE REVIEW

File: Objects/unicodeobject.c [hunk 1]
```diff
@@ -8511,15 +8515,19 @@ encoding_map_lookup(Py_UCS4 c, PyObject *mapping)
    can't be found, Py_None is returned (or NULL, if another
    error occurred). */
 static PyObject *
-charmapencode_lookup(Py_UCS4 c, PyObject *mapping)
+charmapencode_lookup(Py_UCS4 c, PyObject *mapping, unsigned char *replace)
```
Comment (vstinner) [inline Objects/unicodeobject.c:8520]: Can you explain in the comment when replace is set?

File: Objects/unicodeobject.c [hunk 2]
```diff
@@ -8935,15 +8946,20 @@ unicode_translate_call_errorhandler(const char *errors,
    which must be decrefed by the caller.
    Return 0 on success, -1 on error */
 static int
-charmaptranslate_lookup(Py_UCS4 c, PyObject *mapping, PyObject **result)
+charmaptranslate_lookup(Py_UCS4 c, PyObject *mapping, PyObject **result, Py_UCS4 *replace)
```
Comment (vstinner) [inline Objects/unicodeobject.c:8953]: Can you explain in the comment when replace is set?

File: Objects/unicodeobject.c (hunk 1 above)
Comment (serhiy-storchaka) [reply Objects/unicodeobject.c:8520]: Done.

File: Objects/unicodeobject.c [hunk 3]
```diff
@@ -8594,7 +8606,7 @@ charmapencode_output(Py_UCS4 c, PyObject *mapping,
         return enc_SUCCESS;
     }
 
-    rep = charmapencode_lookup(c, mapping);
+    rep = charmapencode_lookup(c, mapping, &replace);
```
Comment (vstinner) [inline Objects/unicodeobject.c:8609]: You can move `unsigned char replace;` declaration here, to reduce the scope of the variable.

**Review by vstinner — APPROVED** [review]
LGTM

File: Objects/unicodeobject.c (hunk 3 above)
Comment (serhiy-storchaka) [reply Objects/unicodeobject.c:8609]: I added it next to the declaration of `rep`, as in all other cases.

**Pull request merged**